### ISSUE TITLE
Make improvements to forward and forward_iter

### ISSUE THREAD

**Issue by benjamin-work** [dnouri/skorch#141]
IMO, [`forward_iter`](https://github.com/dnouri/skorch/blob/9382a83f30d9ba0b1dae5bcc3896feee5e2a0dd5/skorch/net.py#L667) could do with some improvements:

* The docstring "Yield forward results from batches derived from data." is unclear -- it should explain in what conditions to use it.
* The method sets `self.module_.train(training)` but then calls `evaluation_step`, which does the same.

Furthermore, I would adjust [forward](https://github.com/dnouri/skorch/blob/9382a83f30d9ba0b1dae5bcc3896feee5e2a0dd5/skorch/net.py#L728) this way:

```
            y_infer = list(self.forward_iter(X, training=training))
            -->
            y_infer = [yp for yp, *_ in self.forward_iter(X, training=training)]
```

The result of this would be that if `forward_iter` returns more than 1 value (say, like the output of a torch RNN layer), it discards all but the first output. Currently, if more than 1 value is returned, `forward` (and thus `predict` and `predict_proba`) just break, which is discouraging. The change could be a little bit surprising for the user but would make this common use case a lot more comfortable.

**Comment by ottonemo** [comment]
A problem I see is that I would not expect forward to modify the output of my model and it would certainly be helpful if I can rely on the fact that a call to forward on a skorch net will yield the same result as a direct call to my model.

**Comment by benjamin-work** [comment]
I see the problem with the name `forward`. From your proposals, the 2nd makes more sense to me. If we change [this line](https://github.com/dnouri/skorch/blob/9382a83f30d9ba0b1dae5bcc3896feee5e2a0dd5/skorch/net.py#L785) in `predict_proba` from:

`for yp in self.forward_iter(X, training=False):`

to:

`for yp, *_ in self.forward_iter(X, training=False):`

(and the same for `predict`), would that not be enough? I don't see why that would require more memory and neither do I think that we would need a custom collate function. We would just specify that the contract is:

* to use `predict(_proba)`, the first output from `module_.forward` can be cast to numpy and concatenated via `np.concatenate`
* to use `forward`, all outputs must be able to be concatenated by `torch.cat`. (And maybe we need the option to pull batches from GPU to CPU if memory is a concern.)

**Comment by benoua** [comment]
Hi, I had the same issue for a multilabel task. When I compute a recall target at each epoch, the whole predict_proba was not fitting in GPU and I had to do the stacking of results on CPU. 

I used a "to_cpu" option with the modifications below : 
```
    def forward_iter(self, X, training=False, to_cpu=False):
     
        self.module_.train(training)

        dataset = self.get_dataset(X)
        iterator = self.get_iterator(dataset, training=training)
        for Xi, _ in iterator:
            if to_cpu:
                yp = self.evaluation_step(Xi, training=training).cpu()
            else:
                yp = self.evaluation_step(Xi, training=training)
            yield yp
```
Nvertheless, this workaround generated another Variable on gpu, doubling the size of RAM taken for batches:  1 for training batches, 1 for batches required to compute my recall target.

**Comment by BenjaminBossan** [comment]
What you did should not be necessary because `predict_proba` already pulls the data to GPU after each batch via `to_numpy` (see [here](https://github.com/dnouri/skorch/blob/ef863308ff3c699271260b62e9647bec4a5b82ca/skorch/net.py#L781)). Could you please show a minimum code example of computing the recall that results in this memory blow up?

(PS: In case you have an old version of skorch, `predict_proba` could indeed cause memory problems, so updating skorch could already be helpful.)

**Comment by benoua** [comment]
I was using ```forward``` for computing predictions which caused a stacking of probas on GPU : 
```
y_pred = net.forward(X, training=False)
return recall(y_true, y_pred)
```
After your comment, using predict_proba should fix my GPU memory issue instead of using forward. 
I have torch 0.2.0.3

**Comment by benjamin-work** [comment]
For your example, it is indeed better to use `net.predict`, which results in numpy array (CPU). Another advantage is that you can then directly use `sklearn.metrics.recall`.
